oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticSuggestIterator.java

```diff
@@ -0,0 +1,116 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.PriorityQueue;
+
+/**
+ * This class is in charge to extract suggestions for a given query. Suggestion is more like
+ * a completion result.
+ */
+class ElasticSuggestIterator implements Iterator<FulltextResultRow> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticSuggestIterator.class);
+    protected final static String SUGGEST_PREFIX = "suggest?term=";
+
+    private final ElasticIndexNode indexNode;
+    private final ElasticRequestHandler requestHandler;
+    private final ElasticResponseHandler responseHandler;
+    private final String suggestQuery;
+
+    private Iterator<FulltextResultRow> internalIterator;
+    private boolean loaded = false;
+
+    ElasticSuggestIterator(@NotNull ElasticIndexNode indexNode,
+                           @NotNull ElasticRequestHandler requestHandler,
+                           @NotNull ElasticResponseHandler responseHandler) {
+        this.indexNode = indexNode;
+        this.requestHandler = requestHandler;
+        this.responseHandler = responseHandler;
+        this.suggestQuery = requestHandler.getPropertyRestrictionQuery().replace(SUGGEST_PREFIX, "");
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (!loaded) {
+            try {
+                loadSuggestions();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            loaded = true;
+        }
+        return internalIterator != null && internalIterator.hasNext();
+    }
+
+    @Override
+    public FulltextResultRow next() {
+        return internalIterator.next();
+    }
+
+    private void loadSuggestions() throws IOException {
+        BoolQueryBuilder suggestionQuery = requestHandler.suggestionMatchQuery(suggestQuery);
+        final SearchSourceBuilder searchSourceBuilder = SearchSourceBuilder.searchSource()
+                .query(suggestionQuery)
+                .size(100)
+                .fetchSource(FieldNames.PATH, null);
+        final SearchRequest searchRequest = new SearchRequest(indexNode.getDefinition().getRemoteIndexAlias())
+                .source(searchSourceBuilder);
+        SearchResponse res = indexNode.getConnection().getClient().search(searchRequest, RequestOptions.DEFAULT);
+        ArrayList<FulltextResultRow> results = new ArrayList<>();
+        PriorityQueue<ElasticSuggestion> pr = new PriorityQueue<>((a, b) -> Float.compare(b.score, a.score));
+        for (SearchHit doc : res.getHits()) {
+            if (responseHandler.isAccessible(responseHandler.getPath(doc))) {
+                for (SearchHit suggestion : doc.getInnerHits().get(":suggest").getHits()) {
+                    pr.add(new ElasticSuggestion(((List<String>) suggestion.getSourceAsMap().get("suggestion")).get(0), suggestion.getScore()));
```

Comment:
Is there a guarantee that the suggestion map is never empty?

Reply:
this is suspicious indeed

Reply:
IF there is no result it will be stopped at first for loop (res.getHits()). Atleast in current version it is how ES returns result (returns hits only if there is a suggestion/inner_hits).
 If you suggest I can add more empty/null checks.